llm_unlearn_ucl/unlearn_harm.py

```diff
@@ -271,221 +266,72 @@ def main(args) -> None:
-        full_bad_dataset = load_dataset(
-            "sail/symbolic-instruction-tuning", split="train"
-        )
-        if args.shuffle_seed:
-            # shuffle the dataset with a given seed for reproducibility
-            full_bad_dataset = full_bad_dataset.shuffle(seed=args.shuffle_seed)
-        if args.sequential > 0:
-            # NOTE: sequential/batch unlearning using sliced dataset.
-            train_bad_dataset = full_bad_dataset.select(range(args.samples_count))
-        else:
-            # NOTE: full dataset like bytedance.
-            train_bad_dataset = full_bad_dataset
-
-        Path(args.samples_save_dir).mkdir(exist_ok=True)
-        bad_sample_path = f"{args.samples_save_dir}/symbolic_{args.samples_count if args.sequential > 0 else 'full'}_samples.json"
-        with open(bad_sample_path, "w") as fin:
-            print(f"Writing symbolic samples to {bad_sample_path}")
-            json.dump(
-                [
-                    train_bad_dataset[i]
-                    for i in range(
-                        args.samples_count
-                        if args.sequential > 0
-                        else len(train_bad_dataset)
-                    )
-                ],
-                fin,
-            )
-
-        train_bad_loaders = create_symbolic_dataloader_from_dataset(
-            tokenizer,
-            train_bad_dataset,
-            batch_size=args.batch_size,
-            splits=max(args.sequential, 1),
-        )
-
-        question_prefix_str = "### Question:"
-        answer_prefix_str = "### Answer:"
-    elif args.unlearning_dataset == "math_qa":
-        assert (
-            False
-        ), "Mathqa temporarirly disabled - requries implementing returning a List of Datasets for sequential unlearning with equal sizes!"
-        train_dataset = load_dataset("math_qa", split="train")
-        train_bad_loader = create_mathqa_dataloader_from_dataset(
-            tokenizer, train_dataset, batch_size=args.batch_size
-        )
-        question_prefix_str = "Problem:"
-        answer_prefix_str = "rationale:"
     else:
         print(f"Unlearning dataset not known! dataset: {args.unlearning_dataset}")
         return
 
     # Get normal data.
-    if args.retaining_dataset == "truthful_qa":
-        (
-            train_normal_loaders,
-            val_normal_loader,
-            test_normal_loader,
-            train_normal_dataset,
-        ) = create_truthfulqa_dataloader(
-            tokenizer,
-            batch_size=args.batch_size,
-            seed=args.shuffle_seed if args.shuffle_seed is not None else args.seed,
-            num_samples=args.samples_count if args.sequential > 0 else None,
-            splits=max(args.sequential, 1),
+
+    if args.retaining_dataset in SUPPORTED_RETAINING_SET:
+        if args.retaining_dataset == "truthful_qa":
+            args.retaining_dataset = "truthfulqa/truthful_qa"
```

Comment:
why this is a special case?

Reply:
This PR uses huggingface API (`datasets.load_dataset`) to load the truthfulQA dataset (replaces the `csv` file) so that all datasets (mostly) share the same code. The API takes `truthfulqa/truthful_qa` as argument and this special case makes the code backward compatible with existing scripts that uses `truthful_qa` to refer to the dataset.

Reply:
Added comments in the source code for clarity in [8f33ed3](https://github.com/Adamliu1/SNLP_GCW/pull/118/commits/8f33ed3a7549c4828d5df4e6c86addfcfa9180fc).